include/RED4ext/Containers/DynArray.hpp

```diff
@@ -0,0 +1,518 @@
+#include <type_traits>
+
+#include <RED4ext/Common.hpp>
+#include <RED4ext/Detail/AddressHashes.hpp>
+#include <RED4ext/Detail/Containers/ArrayIterator.hpp>
+#include <RED4ext/Memory/Allocators.hpp>
+#include <RED4ext/Relocation.hpp>
+#include <RED4ext/Utils.hpp>
+
+namespace RED4ext
+{
+namespace Memory
+{
+struct IAllocator;
+}
+template<typename T>
+struct DynArray
+{
+    using ValueType = T;
+    using Reference = ValueType&;
+    using ConstReference = const Reference;
+    using Pointer = ValueType*;
+    using ConstPointer = const Pointer;
+
+    using SizeType = std::uint32_t;
+    using DifferenceType = std::ptrdiff_t;
+
+    using Iterator = Detail::ArrayIterator<ValueType, DynArray>;
+    using ConstIterator = Detail::ArrayIterator<const ValueType, DynArray>;
+    using ReverseIterator = std::reverse_iterator<Iterator>;
+    using ConstReverseIterator = std::reverse_iterator<ConstIterator>;
+
+    DynArray(Memory::IAllocator* aAllocator = nullptr)
+        : entries(aAllocator ? std::bit_cast<Pointer>(aAllocator) : nullptr)
+        , size(0)
+        , capacity(0)
+    {
+    }
+
+    DynArray(SizeType aSize, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Resize(aSize);
+    }
+
+    DynArray(std::initializer_list<ValueType> aList, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Assign(aList);
+    }
+
+    DynArray(const DynArray& aOther)
+        : DynArray(aOther.GetAllocator())
+    {
+        Assign(aOther.Begin(), aOther.End());
+    }
+
+    DynArray(DynArray&& aOther) noexcept
+        : entries(aOther.Data())
+        , size(aOther.Size())
+        , capacity(aOther.Capacity())
+    {
+        aOther.entries = std::bit_cast<Pointer>(aOther.GetAllocator());
+        aOther.capacity = 0;
+        aOther.size = 0;
+    }
+
+    template<std::input_iterator InputIt>
+    DynArray(InputIt aFirst, InputIt aLast, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Assign(aFirst, aLast);
+    }
+
+    ~DynArray()
+    {
+        Clear();
+        auto allocator = std::bit_cast<Pointer>(GetAllocator());
+        std::bit_cast<Memory::IAllocator*>(&allocator)->Free(entries);
+        entries = allocator;
+        capacity = 0;
+    }
+
+    DynArray& operator=(const DynArray& aOther)
+    {
+        if (this != std::addressof(aOther))
+        {
+            Assign(aOther.Begin(), aOther.End());
+        }
+        return *this;
+    }
+
+    DynArray& operator=(DynArray&& aOther)
+    {
+        if (this != std::addressof(aOther))
+        {
+            Clear();
+            entries = aOther.Data();
+            size = aOther.Size();
+            capacity = aOther.Capacity();
+
+            aOther.entries = std::bit_cast<Pointer>(aOther.GetAllocator());
+            aOther.capacity = 0;
+            aOther.size = 0;
+        }
+
+        return *this;
+    }
+
+    constexpr Reference operator[](SizeType aPos)
```

Comment:
Why this method and many others are marked as `constexpr`?

Reply:
That's how STL define `std::vector` (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1004r2.pdf) . This should allow the class to be used in `constexpr` functions.

Reply:
But our implementation is not `constexpr`. AFAIK even `std::vector` is still limited to transient allocations inside a constant expression, i.e. it must be destroyed before constant expression evaluation ends. We also won't be able to make our allocators `constexpr` without falling back to `std::allocator` (but it's not worth the effort, you can just use `std::vector` instead, it won't get to runtime anyway).

Reply:
Ah true, I forgot about the allocator being fixed. Then we can just remove it.